### PR TITLE
Explore: Fix how log bars in graph are stacking

### LINKED ISSUE #18796: Explore: Logs bar graph stacking is changing randomly
**What happened**:
When repeating the same query in explore the ordering of the log levels is random and this affects how the bars are stacked, which seems to oscillate between 3 states in which 2 I think are totally wrong:
Screenshot 2019-08-29 at 21 24 14
Screenshot 2019-08-29 at 21 24 20
Screenshot 2019-08-29 at 21 24 26

**What you expected to happen**:
The stacking and color to be consistent.

**How to reproduce it (as minimally and precisely as possible)**:
Using devenv Loki docker container and using query `{job="varlogs"}` seems to do the trick.

**Anything else we need to know?**:
Seems to be related to this in some way https://github.com/grafana/grafana/issues/14247.

**Environment**:
- Grafana version: Grafana v6.4.0-pre (3aa6fdf55)
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
This PR fixes 2 issues: 
1. Logs are being processed in `makeSeriesForLogs` to be displayed in graph. Logs were not sorted and therefore processing was buggy. It was creating additional empty arrays with already used time.
image

2. Time bucket was calculated using `const time = Math.round(row.timeEpochMs / bucketSize) * bucketSize;)`, which sometime resulted in the time, that was in the future, and therefore it wasn't displayed properly on graph. Using `Math.floor()` instead fixes this issue. It will create the same issue on the other side of the graph, but that is probably a better solution. 

The tests were fixed as now the rows are sorted based on` row.timeEpochMs`. 

**Which issue(s) this PR fixes**:
Fixes #18796

### CODE REVIEW

**Review by hugohaggmark — COMMENTED** [review]
Great work with this! I've tested it and it works as expected! I left some small comments for you to consider before I approve.
I found that when you zoom in the graph sometimes, the table has entries but not the graph but that could be related to something completely different.
image

**Comment by aocenas** [comment]
> I found that when you zoom in the graph sometimes, the table has entries but not the graph but that could be related to something completely different.

I think this is because of the interval variable which I think does not change when zooming and is something like 15s by default. In dashboard this is computed automatically from graph width so maybe that part is missing in explore. I think it would be better to do fix it in separate PR though.

File: public/app/core/logs_model.ts [hunk 1]
```diff
@@ -106,7 +106,8 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Grap
   const bucketSize = intervalMs * 10;
   const seriesList: any[] = [];
 
-  for (const row of rows) {
+  const sortedRows = rows.sort((a, b) => (a.timeEpochMs <= b.timeEpochMs ? -1 : 1));
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/core/logs_model.ts]: Btw did not know that you can return 0 in sort function and it will make the sort stable 👍

**Comment by hugohaggmark** [comment]
> > I found that when you zoom in the graph sometimes, the table has entries but not the graph but that could be related to something completely different.
> 
> I think this is because of the interval variable which I think does not change when zooming and is something like 15s by default. In dashboard this is computed automatically from graph width so maybe that part is missing in explore. I think it would be better to do fix it in separate PR though.

I agree

**Comment by ivanahuckova** [comment]
>I think it would be better to do fix it in separate PR though.

I agree! will create an issue for that and I can start to work on it when I am done with this PR. 🙂

File: public/app/core/logs_model.ts [hunk 2]
```diff
@@ -106,7 +107,8 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Grap
   const bucketSize = intervalMs * 10;
   const seriesList: any[] = [];
 
-  for (const row of rows) {
+  const sortedRows = rows.sort(sortInAscendingOrder);
```
Comment (ivanahuckova) [inline public/app/core/logs_model.ts]: @hugohaggmark I have updated this and used the `sortInAscendingOrder` from explore utils, that uses timestamp for sorting. Great point! 🙂

File: public/app/core/logs_model.ts [hunk 3]
```diff
@@ -121,7 +122,7 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Grap
     }
 
     // align time to bucket size
-    const time = Math.round(row.timeEpochMs / bucketSize) * bucketSize;
+    const time = Math.floor(row.timeEpochMs / bucketSize) * bucketSize;
```
[the remark this answers is not in the document]
Comment (ivanahuckova) [reply public/app/core/logs_model.ts]: I think this should be clear and understandable now.

File: public/app/core/specs/logs_model.test.ts [hunk 4]
```diff
@@ -337,18 +337,18 @@ describe('dataFrameToLogsModel', () => {
...
-        logLevel: LogLevel.debug,
-        uniqueLabels: { baz: '1' },
-      },
       {
         entry: 'INFO 1',
         labels: { foo: 'bar', baz: '2' },
         logLevel: LogLevel.error,
         uniqueLabels: { baz: '2' },
       },
+      {
```
[the remark this answers is not in the document]
Comment (ivanahuckova) [reply public/app/core/specs/logs_model.test.ts]: I have added additional test, to test for multiple series with equal times. 🙂

File: public/app/core/logs_model.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/core/logs_model.ts]: Nice!

**Review by hugohaggmark — APPROVED** [review]
LGTM!

**Comment by torkelo** [comment]
> I think this is because of the interval variable which I think does not change when zooming and is something like 15s by default.

That sounds like a bug. Interval should always change when timerange change

**Comment by ivanahuckova** [comment]
@torkelo Yes, it is. I have created a new issue and I am currently working on fixing this: https://github.com/grafana/grafana/issues/19086

**Pull request merged**